kernel/apic.c

```diff
@@ -4,8 +4,107 @@
 #include <stddef.h>
 
 #include "acpi.h"
+#include "com.h"
+#include "hpet.h"
+#include "memory-map.h"
+#include "util.h"
 #include "vga.h"
 
+/* ACPI registers contain 4 bytes of data but use 16 bytes of space.  Writing
+ * to or reading from the upper 12 bytes is always undefined behavior.  Reading
+ * or writing less than an entire 32-bit register at once is also undefined
+ * behavior.  So is using any FP, MMX, SSE etc instruction to read or write
+ * from them.  Be careful. */
+struct APICRegister {
+  uint32_t volatile val;
+  uint32_t unused[3];
+}  __attribute__ ((packed));
+
+/* 1024 bytes long (64 32-bit registers).  They are grouped into sets of 8
+ * registers to make the exact size a little clearer. */
+struct APIC {
+  struct APICRegister reserved_1[2];
+  struct APICRegister local_apic_id;
+  struct APICRegister local_apic_version_id;
+  struct APICRegister reserved_2[4];
+
+  struct APICRegister task_priority;
+  struct APICRegister arbitration_priority;
+  struct APICRegister processor_priority;
+  struct APICRegister end_of_interrupt;
+  struct APICRegister remote_read;
+  struct APICRegister logical_destination;
+  struct APICRegister destination_format;
+  struct APICRegister spurious_interrupt_vector;
+
+  struct APICRegister in_service[8];
+
+  struct APICRegister trigger_mode[8];
+
+  struct APICRegister interrupt_request[8];
+
+  struct APICRegister error_status;
+  struct APICRegister reserved_3[6];
+  struct APICRegister lvt_cmci;
+
+  struct APICRegister interrupt_command[2];
+  struct APICRegister lvt_timer;
+  struct APICRegister lvt_thermal_sensor;
+  struct APICRegister lvt_performance_monitoring_counters;
+  struct APICRegister lvt_lint0;
+  struct APICRegister lvt_lint1;
+  struct APICRegister lvt_error;
+
+  struct APICRegister timer_initial_count;
+  struct APICRegister timer_current_count;
+  struct APICRegister reserved_4[4];
+  struct APICRegister timer_divide_configuration;
+  struct APICRegister reserved_5[1];
+} __attribute__ ((packed));
+
+static const uint32_t APIC_SPURIOUS_VECTOR_ENABLE_APIC = 0x00000100;
+
+struct IOAPIC {
+  uint32_t volatile reg;
+  uint32_t unused[3];
+  uint32_t volatile data;
+} __attribute ((packed));
+
+static const uint32_t IOAPIC_REG_ID             = 0x0000;
+static const uint32_t IOAPIC_REG_VERSION        = 0x0001;
+static const uint32_t IOAPIC_REG_ARBITRATION_ID = 0x0002;
+
+static inline uint32_t IOAPIC_RED_TBL_LOW(uint8_t line) {
+  return 0x10 + line*2;
+}
+static inline uint32_t IOAPIC_RED_TBL_HIGH(uint8_t line) {
+  return 0x10 + line*2 + 1;
+}
+
+static struct APIC *apic;
+static struct IOAPIC *ioapic;
+
+struct isa_irq_remapping {
+  uint8_t isa_line;
+  uint8_t polarity_is_low; /* Normally ISA is active-high */
+  uint8_t trigger_is_level; /* Normally ISA is edge-triggered */
+  uint32_t remapped_apic_line;
+};
+static struct isa_irq_remapping remaps[16];
```

Comment:
what happens when you run out of room for remaps?

Reply:
oh, that's like, hardware impossible isn't it. Is 16 a constant that deserves a name then?

Reply:
Yeah, that's pretty impossible, because there are only 16 ISA lines.  I added a symbolic name, and also added a safety check just in case someone makes an evil MADT that, like, repeatedly remaps the same line and ends up with more than 16 remaps.